docs/project/requirements/architecture.md

```diff
@@ -0,0 +1,147 @@
+# Architecture
+
+Un schéma d'implémentation détaillé sera établi pour décrire l'architecture globale de l'infrastructure, y compris les différents composants, leur interaction et leur déploiement.
+
+
+## Application
+
+
+### Proxy: Traefik
+
+* Application router
+* High-availability by fault tolerance
+* Load-balancing by requests distribution
+  ALB ?
+* Helm chart
+* Version
+
+
+### Application: FastAPI
+
+* Python
+  * Version:
+  * Packaging
+* Project management? Poetry?
+
+
+### Databases: PostgreSQL
+
+* Helm chart
+* Version
+* Master/slave for high availability and read load balancing
+* Master failover
+
+
+## Infrastructure
+
+
+### Network
+
+
+#### Application Load Balancer (ALB)
+
+
+#### Virtual Private Cloud (VPC)
+
+* VPC subnet
+
+
+#### Internet Gateways (IGW)
+
+* Global egress
+
+
+#### Availability Zones (AZ)
+
+* 2 availability zones for high availability
+
+
+#### NAT Gateways (NGW)
+
+* Per AZ egress
+* In public subnet
+
+
+#### Subnets
+
+
+##### Public subnets
+
+* Application access
+* Bastion access
+* NAT gateways ?
+
+
+##### Private subnets
+
+* Database
+* Admin
+
+
+### Compute
+
+#### Nodes
+
+* EC2 compute nodes
+  * Application nodes
+  * Database nodes
+  * Bastion node
+* Auto-scaling groups
+
+
+#### Orchestration: Kubernetes
+
+#### Pods
+
+##### ReplicaSets
+
+* Traefik
+* Application
+* Database
+* cert-manager
+* Metrics server
+* bastion
+
+##### DaemonSets
+
+* Any ?
+
+
+### Storage
+
+#### CSI
+
+##### EBS
+
+* FIXME
+
+
+## CI/CD
+
+### GitHub Actions
+
+
+#### Actions
+
+#### Workflows
+
+* Application
+* Infrastructure
```

Comment:
We have 8 files:
4 workflows are speculative runs that pull information from Terraform Cloud to be added to the pull requests as comments (number of resources deleted, added, modified). These don't many any changes.
2 workflows (preprod) execute terraform apply runs with deployment (automatically executes the changes in the preprod workspace)
2 workflows (prod) execute terraform apply but doesn't deploy, these wait for an operator to approve the change in Terraform Cloud

Reply:
OK, so, conceptually, we really have 4 types of workflows:

- The **provisioning workflows**

  * Build the environments' infrastructure,
  * Run on:
    * each PR and merge for preprod, and
    * each successful prepod for prod, with human validation

- The **test workflow**

  * Tests the application,
  * Runs on each PR and merge before build,
  * Not implemented yet (see preceding comment)

- The **build workflow**

  * Creates application deployment artifacts,
  * Run on each PR and merge now, ideally on each successful test workflow run

- The **deployment workflows**

  * Deploy app artifacts in provisioned environments,
  * Run on:
    * each successful build workflow for preprod, and
    * each successful prepod for prod, with human validation


@JoAngel8 what do you think about this?

Reply:
* The **provisioning workflows**
  * Build the environments' infrastructure,
  * Run on:
    * dry runs on each PR creation and PR change, results posted to PR as comment.
    * each successful merge for preprod, deployed automatically by Terraform Cloud
    * each successful merge for prod, with human validation
 * The **test workflow**
   * Tests the application,
   * Runs on each PR and merge before build,
   * Not implemented yet (see preceding comment)
* The **build workflow**
   * Creates application deployment artifacts,
   * Run on each merge now, ideally on each successful test workflow run
* The **deployment workflows**
   * Deploy app artifacts in provisioned environments,
   * Run on:
     * dry runs on each PR creation and PR change, results posted to PR as comment.
     * each successful merge for preprod, deployed automatically by Terraform Cloud
     * each successful merge for prod, with human validation